core/src/main/java/org/elasticsearch/search/profile/Timer.java

```diff
@@ -33,38 +33,59 @@
  *  }
  *  </pre>
  */
-// TODO: do not time every single call as discussed in https://github.com/elastic/elasticsearch/issues/24799
-public final class Timer {
+public class Timer {
 
-    private long timing, count, start;
+    private boolean doTiming;
+    private long timing, count, lastCount, start;
+
+    /** pkg-private for testing */
+    long nanoTime() {
+        return System.nanoTime();
+    }
 
     /** Start the timer. */
-    public void start() {
+    public final void start() {
         assert start == 0 : "#start call misses a matching #stop call";
+        // We measure the timing of each method call for the first 256
+        // calls, then 1/2 call up to 512 then 1/3 up to 768, etc. with
+        // a maximum interval of 1024, which is reached for 1024*2^8 ~= 262000
+        // This allows to not slow down things too much because of calls
+        // to System.nanoTime() when methods are called millions of time
+        // in tight loops, while still providing useful timings for methods
+        // that are only called a couple times per search execution.
+        doTiming = (count - lastCount) >= Math.min(lastCount >>> 8, 1024);
+        if (doTiming) {
+            start = nanoTime();
+        }
         count++;
-        start = System.nanoTime();
     }
 
     /** Stop the timer. */
-    public void stop() {
-        timing += Math.max(System.nanoTime() - start, 1L);
-        start = 0;
+    public final void stop() {
+        if (doTiming) {
+            timing += (count - lastCount) * Math.max(nanoTime() - start, 1L);
```

Comment:
This was more of a comment than a call-to-action for you. I also do not think that this can be overcome without changing the approach completely. This is the eternal dilemma of tracing vs. sampling. You have to pick one poison. :)